### PR TITLE
feat: DeepSource independent SDK package

### PR DESCRIPTION
**Pull request by siddhant-deepsource**
This pull request makes the following changes :

* Move the API related interaction into an SDK in `./deepsource` folder
* Modify the commands to use the SDK (`deepsource` package) instead of the `api` package
* Implements argument validation for commands
* Some other minor fixes

### CODE REVIEW

File: command/config/validate/validate.go [hunk 1]
```diff
@@ -9,119 +11,77 @@ import (
...
+	"github.com/deepsourcelabs/cli/deepsource/transformers"
+	"github.com/deepsourcelabs/cli/utils"
 	"github.com/pterm/pterm"
 	"github.com/spf13/cobra"
 )
 
 // Options holds the metadata.
 type Options struct {
-	gqlClient *api.DSClient
+	AnalyzerNames      []string
```
Comment (sourya-deepsource) [inline command/config/validate/validate.go:25]: This is repeated. Can it be in the config pkg?

File: command/issues/list/list.go [hunk 2]
```diff
@@ -67,137 +60,131 @@ func NewCmdIssuesList(cf *cmdutils.CLIFactory) *cobra.Command {
 	return cmd
 }
 
+// Execute the command
 func (opts *IssuesListOptions) Run() error {
+	// The current limit of querying issues at once is 100. If the limit passed by user is greater than 100, exit
+	// with an error message
+	if opts.LimitArg > 100 {
```
Comment (sourya-deepsource) [inline command/issues/list/list.go:67]: Hardcoded here. Make it a const.

File: command/issues/list/list.go [hunk 3]
```diff
@@ -67,137 +60,131 @@ func NewCmdIssuesList(cf *cmdutils.CLIFactory) *cobra.Command {
...
-			nodeData = append(nodeData, issueLocation, analyzerShortcode, issueCode, issueTitle)
-			log.Println(nodeData)
+// Gets the data about issues using the SDK based on the user input
+// i.e for a single file or for the whole project
+func (opts *IssuesListOptions) getIssuesData() error {
+	var err error
 
-			opts.ptermTable[index] = nodeData
+	// Get the deepsource client for using the issue fetching SDK to fetch the list of issues
+	deepsource := deepsource.New()
```
Comment (sourya-deepsource) [inline command/issues/list/list.go:146]: Use dependency injection here.

File: command/issues/list/list.go [hunk 4]
```diff
@@ -67,137 +60,131 @@ func NewCmdIssuesList(cf *cmdutils.CLIFactory) *cobra.Command {
...
+	ctx := context.Background()
```
Comment (sourya-deepsource) [inline command/issues/list/list.go:147]: Move this a level up?

File: command/issues/list/list.go [hunk 5]
```diff
@@ -67,137 +60,131 @@ func NewCmdIssuesList(cf *cmdutils.CLIFactory) *cobra.Command {
...
-			nodeData = nil
+	// Case 1 : Fetch issues for a certain FileArg (filepath) passed by the user
+	// Example: `deepsource issues list api/hello.py`
+	if opts.FileArg != "" {
+		// SDK usage for fetching list of issues in a certain file passed by the user
+		opts.issuesData, err = deepsource.GetIssuesForFile(ctx, opts.Owner, opts.RepoName, opts.VCSProvider, opts.FileArg, opts.LimitArg)
+		if err != nil {
+			return err
 		}
+		opts.showIssues(true)
```
Comment (sourya-deepsource) [inline command/issues/list/list.go:157]:
```suggestion
		opts.showIssues(true)
		return nil
```

File: command/issues/list/list.go [hunk 6]
```diff
@@ -67,137 +60,131 @@ func NewCmdIssuesList(cf *cmdutils.CLIFactory) *cobra.Command {
...
 	} else {
+		// Case 2 : Fetch list of issues for the whole project
```
Comment (sourya-deepsource) [inline command/issues/list/list.go:159]: Remove else.

File: command/issues/list/list.go [hunk 7]
```diff
@@ -67,137 +60,131 @@ func NewCmdIssuesList(cf *cmdutils.CLIFactory) *cobra.Command {
...
+		analyzerShortcode := issue.Analyzer.Shortcode
+		issueCode := issue.IssueCode
+		issueTitle := issue.IssueText
 
-			opts.ptermTable[index] = nodeData
+		issueList = append(issueList, issueLocation, analyzerShortcode, issueCode, issueTitle)
 
-			nodeData = nil
-		}
+		opts.ptermTable[index] = issueList
```
Comment (sourya-deepsource) [inline command/issues/list/list.go:185]:
```suggestion
		opts.ptermTable[index] = []string{issueLocation, analyzerShortcode, issueCode, issueTitle}
```

File: command/repo/status/status.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+type RepoStatusOptions struct {
+	RepoArg      string
+	TokenExpired bool
+	Owner        string
+	RepoName     string
+	VCSProvider  string
+}
+
+// NewCmdVersion returns the current version of cli being used
```
Comment (sourya-deepsource) [inline command/repo/status/status.go:23]:
```suggestion
// NewCmdRepoStatus returns the current version of cli being used
```

File: command/repo/status/status.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+		remotesData, err := utils.ListRemotes()
+		if err != nil {
+			if strings.Contains(err.Error(), "exit status 128") {
+				return errors.New("This repository has not been initialized with git. Please initialize it with git using `git init`")
+			}
+		}
+
+		// Condition: If only 1 remote, use it
+		// If more that one remote, give the user choice which one they want to choose
+		if len(remotesData) == 1 {
```
Comment (sourya-deepsource) [inline command/repo/status/status.go:92]: Could we move this to utils?

File: command/repo/view/view.go [hunk 10]
```diff
@@ -55,38 +50,88 @@ func NewCmdRepoView(cf *cmdutils.CLIFactory) *cobra.Command {
...
+		}
+	}
 
-		remotesData, err := cmdutils.ListRemotes()
+	// If the user has access to repo, frame the full URL of the repo and open it on the
+	// default browser
+	var vcs string
+
+	// Extracting VCS name for the URL
+	switch opts.VCSProvider {
```
Comment (sourya-deepsource) [inline command/repo/view/view.go:78]: Could be a map?

File: command/repo/view/view.go [hunk 11]
```diff
@@ -55,38 +50,88 @@ func NewCmdRepoView(cf *cmdutils.CLIFactory) *cobra.Command {
...
+	case "GITHUB":
+		vcs = "gh"
+	case "GITLAB":
+		vcs = "gl"
+	case "BITBUCKET":
+		vcs = "bb"
+	}
+
+	// Framing the complete URL
+	dashboardURL := fmt.Sprintf("https://deepsource.io/%s/%s/%s/", vcs, opts.Owner, opts.RepoName)
```
Comment (sourya-deepsource) [inline command/repo/view/view.go:88]: Hardcoded URL. Will be an issue with on-prem.

File: command/repo/view/view.go [hunk 12]
```diff
@@ -55,38 +50,88 @@ func NewCmdRepoView(cf *cmdutils.CLIFactory) *cobra.Command {
...
+	fmt.Printf("Press enter to open %s in your browser...", dashboardURL)
+	fmt.Scanln()
+	err = browser.OpenURL(dashboardURL)
```
Comment (sourya-deepsource) [inline command/repo/view/view.go:91]:
```suggestion
	return browser.OpenURL(dashboardURL)
```

File: config/config.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+const (
+	ConfigDirName  = "/.deepsource/"
+	ConfigFileName = "/config.toml"
+)
+
+type CLIConfig struct {
+	User                  string
+	Token                 string
+	RefreshToken          string
+	TokenExpiresIn        time.Time
```
Comment (sourya-deepsource) [inline config/config.go:23]: Verify if it contains the timezone info as well.

File: deepsource/auth/mutations/refresh_creds.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+type RefreshTokenParams struct {
+	RefreshToken string `json:"refreshToken"`
+}
+
+type RefreshTokenRequest struct {
+	Params RefreshTokenParams
+}
+
+type RefreshTokenResponse struct {
+	auth.RefreshAuthResponse
```
Comment (sourya-deepsource) [inline deepsource/auth/mutations/refresh_creds.go:30]: Revisit this.

File: deepsource/client.go [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/deepsourcelabs/cli/deepsource/issues"
+	issuesQuery "github.com/deepsourcelabs/cli/deepsource/issues/queries"
+	"github.com/deepsourcelabs/cli/deepsource/repository"
+	repoQuery "github.com/deepsourcelabs/cli/deepsource/repository/queries"
+	"github.com/deepsourcelabs/cli/deepsource/transformers"
+	transformerQuery "github.com/deepsourcelabs/cli/deepsource/transformers/queries"
+	"github.com/deepsourcelabs/graphql"
+)
+
+const host = "https://api.deepsource.io/graphql/"
```
Comment (sourya-deepsource) [inline deepsource/client.go:21]: Read from config?

File: deepsource/issues/queries/list_issues.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	req.Var("name", i.Params.RepoName)
+	req.Var("owner", i.Params.Owner)
+	req.Var("provider", i.Params.Provider)
+	req.Var("limit", i.Params.Limit)
+
+	// set header fields
+	req.Header.Set("Cache-Control", "no-cache")
+	// Adding jwt as header for auth
+	tokenHeader := fmt.Sprintf("JWT %s", client.GetToken())
+	req.Header.Add("Authorization", tokenHeader)
```
Comment (sourya-deepsource) [inline deepsource/issues/queries/list_issues.go:78]: This can be done in a single place?

**Review by sourya-deepsource — CHANGES_REQUESTED** [review]

File: command/config/generate/analyzers_input.go [hunk 17]
```diff
@@ -25,12 +23,11 @@ func (o *Options) collectAnalyzerInput() error {
 
 	// Extracting compulsary meta for Go and Java analyzers
 	for _, analyzer := range o.ActivatedAnalyzers {
-
 		// TODO: Remove this hard coding
```
[the remark this answers is not in the document]
Comment (siddhant-deepsource) [reply command/config/generate/analyzers_input.go:26]: The `required` field is missing from the `meta_schema` of Go and Java analyzers. Both have only `optional_required` field. Should I input the `optional_required` fields instead?

[the remark this answers is not in the document]
Comment (siddhant-deepsource) [reply command/config/generate/analyzers_input.go:26]: Will address this in a separate PR.

**Review by vishnu-deepsource — APPROVED** [review]

File: global/global.go [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package global
+
+var (
+	Host         string = "https://api.deepsource.io/graphql"
+	User         string
+	Token        string
+	RefreshToken string
+	TokenExpired bool = true
+)
```
[the remark this answers is not in the document]
Comment (siddhant-deepsource) [reply global/global.go]: Had added this package earlier. Deleted in further commits to use the `config` package itself.

**Pull request merged**